src/System.Linq/src/System/Linq/Where.cs

```diff
@@ -243,18 +313,51 @@ public override IEnumerable<TResult> Select<TResult>(Func<TSource, TResult> sele
                 return new WhereSelectListIterator<TSource, TResult>(_source, _predicate, selector);
             }
 
+            public TSource[] ToArray()
+            {
+                var builder = new LargeArrayBuilder<TSource>(initialize: true);
+
+                List<TSource> source = _source;
+                for (int i = 0; i < source.Count; i++)
+                {
+                    TSource item = source[i];
```

Comment:
For Lists, using `foreach` will be slower since the compiler doesn't optimize that.